### PR TITLE
Update fc_core.c

### PR DESCRIPTION
**Pull request by KenImhof**
Addresses Issue #2120 With 3D mode arming not working properly.  I also suspect that this bug could address other arming issues as the logic for checking the arming flag was inverted.

### CODE REVIEW

File: src/main/fc/fc_core.c [hunk 1]
```diff
@@ -238,9 +238,9 @@ static void updateArmingStatus(void)
         /* CHECK: Arming switch */
         if (!isUsingSticksForArming()) {
             // If arming is disabled and the ARM switch is on
-            if (isArmingDisabled() && IS_RC_MODE_ACTIVE(BOXARM)) {
+            if (isArmingDisabled() && !IS_RC_MODE_ACTIVE(BOXARM)) {     //kbi Issue #2120 3D Mode Arming Not Working
```
Comment (digitalentity) [inline src/main/fc/fc_core.c]: This is incorrect. If arming was disabled before and arming switch is "ON" we need to keep arming disabled.

File: src/main/fc/fc_core.c [hunk 2]
```diff
@@ -238,9 +238,9 @@ static void updateArmingStatus(void)
...
                 ENABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);
-            } else if (!IS_RC_MODE_ACTIVE(BOXARM)) {
+            } else if (IS_RC_MODE_ACTIVE(BOXARM)) {                     //kbi Issue #2120 3D Mode Arming Not Working
```
Comment (digitalentity) [inline src/main/fc/fc_core.c]: Similarly, if ARM mode is not active - disable the "ARMING_DISABLED" flag

**Review by digitalentity — COMMENTED** [review]
This change breaks the logic. Please explain the change.

**Comment by KenImhof** [comment]
Hi Konstantin,

 

The problem is that if arming occurs when (THROTTLE != LOW) With SWITCH Armed then we have:

       isArmingDisabled = TRUE and 

IS_RC_MODE_ACTIVE(BOXARM) == TRUE

 

So we set the ENABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);

 

 

When the throttle is fixed (THROTTLE == LOW):

       isArmingDisabled == TRUE due to the ARMING_DISABLED_ARM_SWITCH

IS_RC_MODE_ACTIVE(BOXARM) == TRUE (Switch Still Armed)

So we continue to set the ENABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH) even though the switch is in the arm position.

 

Problem is we never ARM even though THROTTLE == LOW With Switch Armed

 

I have verified my code fix is working, but perhaps a better solution might be to check if the only enabled disarm flag set is the disabled_arm_switch, and the Switch is in the armed position, then remove the disabled_arm_switch flag.

**Comment by KenImhof** [comment]
This is probably more like the original 1.7.2 functionality that is needed.  If the only disarm flag set is the throttle when the arm switch is enabled, then the FC will arm when the throttle is corrected.  I also added code to check if any other disarm flags are set when the arm switch is enabled and it will not arm until the flags are cleared and the arm switch is toggled.

 

Please let me know if I should submit a pull request.  The update code is as follows:

 

static void updateArmingStatus(void)

{

 

    if (ARMING_FLAG(ARMED)) {

        LED0_ON;

    } else {

        /* CHECK: Run-time calibration */

        static bool calibratingFinishedBeep = false;

        if (isCalibrating()) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_SENSORS_CALIBRATING);

            calibratingFinishedBeep = false;

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_SENSORS_CALIBRATING);

 

            if (!calibratingFinishedBeep) {

                calibratingFinishedBeep = true;

                beeper(BEEPER_RUNTIME_CALIBRATION_DONE);

            }

        }

 

        /* CHECK: RX signal */

        if (!failsafeIsReceivingRxData()) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_RC_LINK);

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_RC_LINK);

        }

 

        /* CHECK: Throttle */

        if (calculateThrottleStatus() != THROTTLE_LOW) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_THROTTLE);

        }

        else if (armingFlags & ARMING_DISABLED_THROTTLE ){ //Previous Throttle Not Low - kbi

                DISABLE_ARMING_FLAG(ARMING_DISABLED_THROTTLE);

                if ( (armingFlags & ARMING_DISABLED_ALL_FLAGS)==ARMING_DISABLED_ARM_SWITCH && IS_RC_MODE_ACTIVE(BOXARM)  ){

                    DISABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);  //Arm Now That Throttle is Fixed and No Other Flags Set

                }

         }

        else{

            DISABLE_ARMING_FLAG(ARMING_DISABLED_THROTTLE);

        }

 

        /* CHECK: Angle */

        if (!STATE(SMALL_ANGLE)) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_NOT_LEVEL);

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_NOT_LEVEL);

        }

 

        /* CHECK: CPU load */

        if (isSystemOverloaded()) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_SYSTEM_OVERLOADED);

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_SYSTEM_OVERLOADED);

        }

        

#if defined(NAV)

        /* CHECK: Navigation safety */

        if (navigationBlockArming()) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_NAVIGATION_UNSAFE);

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_NAVIGATION_UNSAFE);

        }

#endif

 

#if defined(MAG)

        /* CHECK: */

        if (sensors(SENSOR_MAG) && !STATE(COMPASS_CALIBRATED)) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_COMPASS_NOT_CALIBRATED);

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_COMPASS_NOT_CALIBRATED);

        }

#endif

 

        /* CHECK: */

        if (sensors(SENSOR_ACC) && !STATE(ACCELEROMETER_CALIBRATED)) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_ACCELEROMETER_NOT_CALIBRATED);

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_ACCELEROMETER_NOT_CALIBRATED);

        }

 

        /* CHECK: */

        if (!isHardwareHealthy()) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_HARDWARE_FAILURE);

        }        

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_HARDWARE_FAILURE);

        }

 

        /* CHECK: Arming switch */

        if (!isUsingSticksForArming()) {

            static bool ArmSwitchToggle=false;

            // If ARM switch is Off or Other Disarm Flags Set

            if ( isArmingDisabled() || !IS_RC_MODE_ACTIVE(BOXARM) ) { //kbi

              ENABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);

              if (!IS_RC_MODE_ACTIVE(BOXARM)){

                  ArmSwitchToggle = true;

              }

              else if (ArmSwitchToggle && (armingFlags & ARMING_DISABLED_ALL_FLAGS) == ARMING_DISABLED_ARM_SWITCH ){

                  DISABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);

                  ArmSwitchToggle=false;

              }

              else{

                  ArmSwitchToggle=false;

              }

 

            }

            else if (ArmSwitchToggle){

                DISABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);

                ArmSwitchToggle=false;

            }

        }

 

        /* CHECK: BOXFAILSAFE */

        if (IS_RC_MODE_ACTIVE(BOXFAILSAFE)) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_BOXFAILSAFE);

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_BOXFAILSAFE);

        }

 

        /* CHECK: BOXFAILSAFE */

        if (IS_RC_MODE_ACTIVE(BOXKILLSWITCH)) {

            ENABLE_ARMING_FLAG(ARMING_DISABLED_BOXKILLSWITCH);

        }

        else {

            DISABLE_ARMING_FLAG(ARMING_DISABLED_BOXKILLSWITCH);

        }

 

        if (isArmingDisabled()) {

            warningLedFlash();

        } else {

            warningLedDisable();

        }

 

        warningLedUpdate();

    }

}

 

 

From: Ken [mailto:KenImhof@roadrunner.com] 
Sent: Tuesday, October 31, 2017 9:28 AM
To: 'iNavFlight/inav' <reply@reply.github.com>; 'iNavFlight/inav' <inav@noreply.github.com>
Cc: 'Author' <author@noreply.github.com>
Subject: RE: [iNavFlight/inav] Update fc_core.c (#2433)

 

Hi Konstantin,

 

The problem is that if arming occurs when (THROTTLE != LOW) With SWITCH Armed then we have:

       isArmingDisabled = TRUE and 

IS_RC_MODE_ACTIVE(BOXARM) == TRUE

 

So we set the ENABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);

 

 

When the throttle is fixed (THROTTLE == LOW):

       isArmingDisabled == TRUE due to the ARMING_DISABLED_ARM_SWITCH

IS_RC_MODE_ACTIVE(BOXARM) == TRUE (Switch Still Armed)

 

So we continue to set the ENABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH) even though the switch is in the arm position.

 

Problem is we never ARM even though THROTTLE == LOW With Switch Armed

 

I have verified my code fix is working, but perhaps a better solution might be to check if the only enabled disarm flag set is the disabled_arm_switch, and the Switch is in the armed position, then remove the disabled_arm_switch flag.

File: src/main/fc/fc_core.c [hunk 3]
```diff
@@ -237,11 +245,25 @@ static void updateArmingStatus(void)
...
         /* CHECK: Arming switch */
         if (!isUsingSticksForArming()) {
-            // If arming is disabled and the ARM switch is on
-            if (isArmingDisabled() && IS_RC_MODE_ACTIVE(BOXARM)) {
-                ENABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);
-            } else if (!IS_RC_MODE_ACTIVE(BOXARM)) {
+            static bool ArmSwitchToggle=false;
+            // If ARM switch is Off or Other Disarm Flags Set
+            if ( isArmingDisabled() || !IS_RC_MODE_ACTIVE(BOXARM) ) { //kbi
+              ENABLE_ARMING_FLAG(ARMING_DISABLED_ARM_SWITCH);
```
Comment (digitalentity) [inline src/main/fc/fc_core.c:251]: I don't understand this logic. The `isArmingDisabled() || !IS_RC_MODE_ACTIVE(BOXARM)` check will disable arming for "ARM SWITCH" reason when the switch is in fact not activated.

**Comment by digitalentity** [comment]
I think the desired behavior can be acheived by not treating full reverse thrust in 3D mode as arming blocker.

**Comment by KenImhof** [comment]
I agree and this is a much simpler method.  I will test out this approach today.

**Pull request closed without merging**